### ISSUE TITLE
softrealtime option will only apply to main thread of the client's process

### ISSUE THREAD

**Issue by mdiluz** [FeralInteractive/gamemode#139]
**Describe the bug**
When using the `softrealtime` option, the gamemode daemon calls `sched_setscheduler` here: https://github.com/FeralInteractive/gamemode/blob/master/daemon/gamemode-sched.c#L118.

The man page explains that these are per-thread attributes, so only the main thread of the client will have it's scheduler set. 

>        The scheduling policy and parameters are in fact per-thread
>       attributes on Linux.  The value returned from a call to gettid(2) can
>       be passed in the argument pid.  Specifying pid as 0 will operate on
>       the attributes of the calling thread, and passing the value returned
>       from a call to getpid(2) will operate on the attributes of the main
>       thread of the thread group.  (If you are using the POSIX threads API,
>       then use pthread_setschedparam(3), pthread_getschedparam(3), and
>       pthread_setschedprio(3), instead of the sched_*(2) system calls.)

This may not be the intended behavior, as the user would likely want `SCHED_ISO` to be set on threads such as the graphics dispatch thread.

**To Reproduce**
Steps used to reproduce the behavior:
1. Have a `-ck` kernel installed
2. Configure `gamemode.ini` with `softrealtime=on`
2. Request gamemoded from a multithreaded program
3. Call `sched_getscheduler` on each thread and see the result

**Expected behavior**
All threads should have `SCHED_ISO` set.

**System Info (please complete the following information):**
 - System with `-ck` kernel
 - 1.3.1

**Additional context**
*NOTE:* I have not been able to verify this on a `-ck` kernel yet, but if the man page above is to be believed, the bug exists.

**Comment by mdiluz** [comment]
This will need the equivalent of the https://github.com/FeralInteractive/gamemode/pull/142/commits/0fb9cbd4539f44e06b84053c54e555978f9a8c12 change, assuming that the man page isn't lying about `The scheduling policy and parameters are in fact per-thread attributes on Linux`. It'll also need tests implemented similar to those in #142.

However, I'm not 100% sure that this can operate in the same way without needing sudo permissions - multiple places online (like [on the ck kernel wiki](https://ck.fandom.com/wiki/SchedulingPolicies)) explain that `SCHED_ISO` is not simple to remove without root.

@kakra I assume you'd know how to fix up the above?

**Comment by kakra** [comment]
I don't think this needs fixing. It is expected that a game supporting native GameMode calls GameMode from a parent thread.

Also, it is very important that scheduler settings never leak into children processes, otherwise this calls for very unexpected behavior. There was a bug report about this a while ago.

What about games that don't have native support for GameMode? After all, they may run through a launcher or start script... Well, in that case, `libgamemodeauto.so` will be preloaded via environment variables. This variable will be carried over to child processes unless something explicitly overwrites or unsets the variable.

The other thing to note is: In Linux, threads and processes aren't really different. Linux has no understanding of the difference of both, thread ids are in fact process ids. The difference is only when spawning new threads: Memory is shared differently. This is why the man page explicitly mentions that scheduling parameters are per thread: Because each new thread is a process from kernel view, and processes don't share (but inherit) scheduling settings.

It is true, that once `SCHED_ISO` is set, it cannot be unset. That's why (A) it must not leak into children, and (B) we don't need to save a previous state: The setting just dies when the process ends.

I'm not sure if thread creation vs process forking would inherit the settings into children. I didn't really test this but instead was hoping that games supporting native GameMode would call it from the correct threads.

If you want to apply settings to the whole group of processes/threads, we should probably use the process group id. Children should share the same PGID, and almost all syscalls support setting values also by PGID, not only by PID. This doesn't fix inheriting settings, tho. I'm not sure how rtkit solves this. It probably expects each thread to acquire RT via an API call. GameMode should probably support something similar through its DBUS API.

**Comment by mdiluz** [comment]
So to clear something up - GameMode should be able to be applied and removed at any time, requested by the process itself, or a "supervisor" process. If an optimization is only applied to the main thread when an external "supervisor" requests GameMode for an already running process, then that should be considered a bug, and the same applies for when an optimization isn't removed fully whenever the opposite request comes in.

It's important because by ensuring we stick to that behavior, we keep GameMode conceptually able to be used in a far more automated way than it currently is.

That said though:

> It is true, that once SCHED_ISO is set, it cannot be unset. That's why (A) it must not leak into children, and (B) we don't need to save a previous state: The setting just dies when the process ends.

This is a really good point. It functionally breaks the above, so I think it needs some real thought. One could say that GameMode isn't the place for that kind of optimization, but I'd much rather we work out a way to make GameMode properly understand optimizations that match this pattern and deal with them elegantly.

> If you want to apply settings to the whole group of processes/threads, we should probably use the process group id. Children should share the same PGID, and almost all syscalls support setting values also by PGID, not only by PID.

Yes, PGID usage should get a second look. I _feel_ like it didn't work for me when I was writing the thread walking code but I could have just been getting it wrong.

>  This doesn't fix inheriting settings, tho. I'm not sure how rtkit solves this. It probably expects each thread to acquire RT via an API call. GameMode should probably support something similar through its DBUS API.

Agreed, something thread specific seems like the best way to solve this. And yeah RTKit is used per-thread, not per-process. In an ideal world GameMode could externally work out which threads are actually in need of scheduling optimizations, but that might be very hard to get right.

**Comment by kakra** [comment]
BTW: `SCHED_ISO` can be removed with proper `CAP_SYS_*` privileges, or by running with root privileges. You could make this into a helper as you did for `cpugovctl`. But GameMode itself won't be able to do this internally as it runs with only user privileges.

The problem with leaking into child processes is that such processes could no longer manage their own niceness when `SCHED_ISO` leaked into them. I'm not sure if ioprio has similar restrictions. I guess it's the same as with nice: A user-privileged process can lower its priority but cannot raise it again. `SCHED_ISO` is an extraordinary exception here: It works quite the other way round: A process can enter `SCHED_ISO` mode but not leave it anymore because that's considered a raising priority by the kernel - maybe because of the class values (`SCHED_OTHER` = 3 vs. `SCHED_ISO` = 4).

The bug I mentioned was about a resource manager (similar to GameMode) leaking `SCHED_IDLE` priority of the Steam launcher into games launched by Steam. The game or GameMode was no longer able to reset `SCHED_IDLE` to something else. Conclusion: Don't manage launchers with resource managers. In consequence, games must not leak their GameMode settings into children processes. Imagine a game launching your browser, and that has a lot of tabs open with CPU hungry javascripts. The result would probably not be that funny.
